airrohr-firmware/mhz19.cpp

```diff
@@ -0,0 +1,150 @@
+ /***********************************************************************
+ *    MH-Z19 CO2 sensor driver                                          *
+ *    Copyright (C) 2017-2019 Michael Dreher <michael (at) 5dot1.de>    *
+ *                                                                      *
+ * This program is free software: you can redistribute it and/or modify *
+ * it under the terms of the GNU General Public License as published by *
+ * the Free Software Foundation, either version 3 of the License, or    *
+ * (at your option) any later version.                                  *
+ *                                                                      *
+ * This program is distributed in the hope that it will be useful,      *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of       *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        *
+ * GNU General Public License for more details.                         *
+ *                                                                      *
+ * You should have received a copy of the GNU General Public License    *
+ * along with this program. If not, see <http://www.gnu.org/licenses/>. *
+ ***********************************************************************/
+
+#include "mhz19.h"
+
+CMHZ19Sensor::CMHZ19Sensor(Stream& serialPort) :
+	m_pSerialPort(&serialPort),
+	ppmMaxValue(0xffff)
+{
+}
+
+/*
+set the measuring range to 1000, 2000, 3000 or 5000
+ppm range for sensor MH-Z19 is set to 2000 when sensor is delivered
+*/
+bool CMHZ19Sensor::SetRange(uint16_t range)
+{
+  if((range % 1000) == 0
+    && range >= 1000
+    && range <= 5000
+  )
+  {
+    uint8_t cmd[packetLen] = {0xFF, 0x01, 0x99, 0x00, 0x00, 0x00, (uint8_t)((range >> 8) & 0xff), (uint8_t)(range & 0xff), 0x00}; // 2nd uint8_t is the sensor number
+    cmd[packetLen - 1] = CalcCheckSum(cmd);
+    m_pSerialPort->write(cmd, sizeof(cmd));
+    // TODO: check return value if any
+    // consume any response in the receive buffer
+    delay(200);
+    while(m_pSerialPort->available())
+      m_pSerialPort->read();
+
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+uint16_t CMHZ19Sensor::GetRange()
+{
+  return ppmMaxValue;
+}
+
+
+// returns 0xffff when reading fails
+uint16_t CMHZ19Sensor::ReadCO2Sensor(){
+  uint16_t ppm = 0xffff;
+  uint8_t response[packetLen]; 
+  static const uint8_t cmd[packetLen] = {0xFF,0x01,0x86,0x00,0x00,0x00,0x00,0x00,0x79}; // 2nd uint8_t is the sensor number
+  m_pSerialPort->write(cmd, sizeof(cmd));
+  if(!SynchronizeStream())
+  {
+#if DEBUG_OUTPUT    
+      Serial.print(F("Stream not synchronized "));
+#endif
+  }
+  else
+  {
+    uint8_t recvCnt = m_pSerialPort->readBytes((char*)response, sizeof(response));
+
+#if DEBUG_OUTPUT    
+    for(int8_t i = 0; i < min(recvCnt, packetLen); i++)
+    {
+      Serial.print(response[i], HEX);
+      Serial.write(' ');
+    }
+#endif
```

Comment:
I believe some of these lines have trailing whitespace (it somehow isn't visible in githubs view)

Reply:
removed all whitespaces at EOL from all source files